cli/cache/http_cache/local.rs

```diff
@@ -0,0 +1,564 @@
+use super::HttpCache;
+use super::HttpCacheItemKey;
+
+#[derive(Debug)]
+pub struct LocalHttpCache {
+  path: PathBuf,
+  manifest: LocalCacheManifest,
+  global_cache: Arc<GlobalHttpCache>,
+}
+
+impl LocalHttpCache {
+  pub fn new(path: PathBuf, global_cache: Arc<GlobalHttpCache>) -> Self {
+    assert!(path.is_absolute());
+    let manifest = LocalCacheManifest::new(path.join("manifest.json"));
+    Self {
+      path,
+      manifest,
+      global_cache,
+    }
+  }
+
+  fn get_cache_filepath(
+    &self,
+    url: &Url,
+    headers: &HeadersMap,
+  ) -> Result<PathBuf, AnyError> {
+    Ok(url_to_local_sub_path(url, headers)?.as_path_from_root(&self.path))
+  }
+
+  /// Copies the file from the global cache to the local cache returning
+  /// if the data was successfully copied to the local cache.
+  fn check_copy_global_to_local(&self, url: &Url) -> Result<bool, AnyError> {
+    let global_key = self.global_cache.cache_item_key(url)?;
+    let Some(metadata) = self.global_cache.read_metadata(&global_key)? else {
+      return Ok(false);
+    };
+
+    if !metadata.is_redirect() {
+      let Some(cached_bytes) = self.global_cache.read_file_bytes(&global_key)? else {
+        return Ok(false);
+      };
+
+      let local_file_path = self.get_cache_filepath(url, &metadata.headers)?;
+      // if we're here, then this will be set
+      ensure_dir_exists(local_file_path.parent().unwrap())?;
+      atomic_write_file(&local_file_path, cached_bytes, CACHE_PERM)
+        .with_context(|| {
+          format!("Failed saving '{}'", local_file_path.display())
+        })?;
+    }
+    self.manifest.insert_data(
+      url_to_local_sub_path(url, &metadata.headers)?,
+      url.clone(),
+      metadata.headers,
+    );
+
+    Ok(true)
+  }
+
+  fn get_url_metadata_checking_global_cache(
+    &self,
+    url: &Url,
+  ) -> Result<Option<CachedUrlMetadata>, AnyError> {
+    if let Some(metadata) = self.manifest.get_metadata(url) {
+      Ok(Some(metadata))
+    } else if self.check_copy_global_to_local(url)? {
+      // try again now that it's saved
+      Ok(self.manifest.get_metadata(url))
+    } else {
+      Ok(None)
+    }
+  }
+}
+
+impl HttpCache for LocalHttpCache {
+  fn cache_item_key<'a>(
+    &self,
+    url: &'a Url,
+  ) -> Result<HttpCacheItemKey<'a>, AnyError> {
+    Ok(HttpCacheItemKey {
+      #[cfg(debug_assertions)]
+      is_local_key: true,
+      url,
+      file_path: None, // need to compute this every time
+    })
+  }
+
+  fn contains(&self, url: &Url) -> bool {
+    self.manifest.get_metadata(url).is_some()
+  }
+
+  fn read_modified_time(
+    &self,
+    key: &HttpCacheItemKey,
+  ) -> Result<Option<SystemTime>, AnyError> {
+    debug_assert!(key.is_local_key);
+    self
+      .get_url_metadata_checking_global_cache(key.url)
+      .map(|m| m.map(|m| m.time))
+  }
+
+  fn set(
+    &self,
+    url: &Url,
+    headers: crate::http_util::HeadersMap,
+    content: &[u8],
+  ) -> Result<(), AnyError> {
```

Comment:
Is this function called on each run or is there some smart caching going on to limit the work?

Reply:
It's only called when the url wasn't in the cache. I just added a print statement here and ran it to make sure.

Reply:
Sounds good to me!